doc/logging/persistent_logger/persistent_loglevel.md

```diff
@@ -0,0 +1,385 @@
+This change will impact the following components:
+
+buffermgrd                    
+coppmgrd                      
+fdbsyncd                      
+fpmsyncd                      
+gearsyncd                     
+intfmgrd                     
+nbrmgrd
+neighsyncd
+orchagent
+portmgrd                      
+portsyncd                  
+syncd                 
+teammgrd                  
+teamsyncd               
+tlm_teamd                    
+tunnelmgrd                 
+vlanmgrd                      
+vrfmgrd                      
+vxlanmgrd                    
+wjhd   
+
+The change is the file: /sonic-swss-common/common/logger.cpp
+
+## 3.2 Persistent Logger flow
+
+- Each component has a singleton Logger object with a log level property and listener thread.
+- When a component writes a log message, the Logger writes the message only if the loglevel of the message is above the current loglevel property.
+- When the user wants to set a new log level to a component, he uses the "swssloglevel" script. The "swssloglevel" script sets the new verbosity to the CONFIG DB (database #4).
+- The CONFIG DB change triggers an event, which is caught by the listener thread.
+- The listener thread changes the loglevel property of the Logger.
+- The user can use the "config save" command to save the current loglevel and make it persistent. It will copy the CONFIG DB content to the config_db.json.
+
+In addition to the log level, the CONFIG DB contains the log output file. After "config save" the log output will be persistent to reboot too.
+
+
+
+
+
+
+
+![persistent logger flow](/doc/logging/persistent_logger/persistent_logger.png)
+
+
+
+
+
+
+
+# 4 Flows
+
+## 4.1 System init flow
+
+When the system startup and the Database container initialize, and the config_db.json loads to CONFIG DB.
+ 
+  - There is a concern about the log level of messages written before the database container is initialized (if it exists). From my understanding, the RSYSLOG-CONFIG is up only after the DATABASE container, so there won't be logs before the config_db.json loads to the CONFIG DB. In case there are messages before the loading, the messages will be in the default log level.
+
+
+# 5 Cold and Fast Reboot Support
+  
+  The current implementation support cold and fast reboot. Sine in the cold and fast reboot, the CONFIG DB content is deleted, the user needs to run "config save" to make the log level persistent to cold and fast reboot.
+
+# 6 Warm Reboot Support
+
+  With the current implementation, we don't flush the CONFIG DB before warm-reboot, which means that if the user configures some loglevel (for example, debug), after warm-reboot, the system startup with the same configurable loglevel (debug).
```

Comment:
>implementation [](http://example.com/codeflow?start=19&length=14)

Seems you only consider warm-reboot into the same image. How about warm upgrade and warm downgrade into another image, which has different ConfigDB schema? I am thinking you need add logic into db_migrator.

Reply:
Also add test cases on warm upgrade/downgrade.